Optimise division by a constant at runtime for integer division

JAicewizard opened the pull request:
When the rhs of an integer divide is known to be constant, it is possible to optimize this. Libdivide does this optimization at runtime, so even when the constant isn't known at compile time, it can still perform similar optimizations.

As you can see below the code using libdivide runs up to twice as fast. It should be even faster for unsigned integers.
It might be possible for the compiler to auto-vectorise the branchless version of libdivide, but I have not looked into that. Open for future research.

## Issues with this code

I personally find the current method a bit messy, as it is the responsibility of the OPWRAPPER to decide what to optimize and what not. On one hand this allows the OP to be as generic as it was before. On the other hand, adding this libdivide optimization to more types will result in lots of duplicate code across the wrappers.

## Benchmarks (ran with turbo-boost disabled)

No optimisation:
```
name	run	timing
benchmark/micro/arithmetic/division_constrhs.benchmark	1	1.170270
benchmark/micro/arithmetic/division_constrhs.benchmark	2	1.173304
benchmark/micro/arithmetic/division_constrhs.benchmark	3	1.164200
benchmark/micro/arithmetic/division_constrhs.benchmark	4	1.179625
benchmark/micro/arithmetic/division_constrhs.benchmark	5	1.177252
```

Just performing row validity check based on the rhs (if possible):
```
name	run	timing
benchmark/micro/arithmetic/division_constrhs.benchmark	1	1.113453
benchmark/micro/arithmetic/division_constrhs.benchmark	2	1.130460
benchmark/micro/arithmetic/division_constrhs.benchmark	3	1.125331
benchmark/micro/arithmetic/division_constrhs.benchmark	4	1.134362
benchmark/micro/arithmetic/division_constrhs.benchmark	5	1.120623
```

Using libdivide:
```
name	run	timing
benchmark/micro/arithmetic/division_constrhs.benchmark	1	0.498989
benchmark/micro/arithmetic/division_constrhs.benchmark	2	0.497163
benchmark/micro/arithmetic/division_constrhs.benchmark	3	0.498707
benchmark/micro/arithmetic/division_constrhs.benchmark	4	0.504418
benchmark/micro/arithmetic/division_constrhs.benchmark	5	0.502800
```

JAicewizard commented:
Builds are all failing seemingly because of formatting issues.
As mentioned the biggest issue I have with the code is that the responsibility of optimization is with the wrapper not the operation itself. As I am not very familiar with c++ templates, I don't really know how to improve this.
However if this isn't an issue on your side, I can remove the WIP and fix the formatting issues.

Similar optimisations can be done for the modulo operator. I will file a seperate PR for that once this is merged.

Tagging @lnkuiper since I already mentioned this to him on monday

Mytherin reviewed (commented):
Thanks for the PR! Great performance results.

* Can we create a separate `divide_by_const` function, and then rewrite `x // C` to `divide_by_const(x, C)` as an optimization in the `ArithmeticSimplificationRule`, instead of doing this optimization within the division operator?
* We avoid explicit SIMD instructions in DuckDB - and `libdivide` seems to contain many of them. In general it seems like a very complex library for what should be a relatively simple optimization. I wonder if we could either strip down libdivide, or switch to something like [fastmod](https://github.com/lemire/fastmod) which seems to be a lot more simple.

JAicewizard commented:
> Can we create a separate divide_by_const function, and then rewrite x // C to divide_by_const(x, C) as an optimization in the ArithmeticSimplificationRule, instead of doing this optimization within the division operator?

I don't know! I don't know a lot of the internals of duckdb. I can write a function that does this, but I am not sure I can also do the optimization part of it. This does however seam like a much cleaner solution.

> We avoid explicit SIMD instructions in DuckDB - and libdivide seems to contain many of them. In general it seems like a very complex library for what should be a relatively simple optimization. I wonder if we could either strip down libdivide, or switch to something like [fastmod](https://github.com/lemire/fastmod) which seems to be a lot more simple.

I havn't yet tested fastmod, that was the library I was intending to use for the follow-up using modulo. One advantage  of this library is that it also provides branchless versions, which may prove advantageous for automatic vectorisation. I also don't know the performance of fastmod for division
All the explicit vectorisation can be removed I think, I will look into that.

Mytherin commented:
> > Can we create a separate divide_by_const function, and then rewrite x // C to divide_by_const(x, C) as an optimization in the ArithmeticSimplificationRule, instead of doing this optimization within the division operator?
> 
> I don't know! I don't know a lot of the internals of duckdb. I can write a function that does this, but I am not sure I can also do the optimization part of it. This does however seam like a much cleaner solution.
> 

Have a look [here](https://github.com/duckdb/duckdb/blob/7a166b22247d81407195f55e107803d8aceb0129/src/optimizer/rule/arithmetic_simplification.cpp#L59), I think the rewrite should be relatively straightforward.

JAicewizard commented:
I implemented it as a function instead of an operator. I wasn't entirely sure where to put it, but it can easily be moved of course.

I also looked into using `fastmod` for division, however it only supports 32 and 64 bit integers, and doesnt support 64 bit division on MSVC at all. I also measured the performance, and it is significantly slower.

xuke-hat commented on src/core_functions/scalar/math/numeric.cpp:
```diff
@@ -1320,4 +1326,88 @@ ScalarFunctionSet LeastCommonMultipleFun::GetFunctions() {
+
+	ValidityMask &validity = FlatVector::Validity(result);
+	if (rhs == 0) {
+		validity.SetInvalid(args.data.size());
```
`SetInvalid` sets the given index to null, I guess you want to use `SetAllInvalid`? 
And may be it is better to set result a constant null vector:
```
result.SetVectorType(VectorType::CONSTANT_VECTOR);
ConstantVector::SetNull(result, true);
```

JAicewizard replied:
Ah I don't know why I didn't notice I wasn't using `SetAllInvalid`. But I changed it to return a constant null vector anyways which is probably better

JAicewizard commented:
I moved this PR to use fastmod. This reduced the available types this optimization applies to, to `uint32_t`, but performance is around the same.

JAicewizard commented:
I reimplemented fastmod using the duckdb 128 bit types to allow unsigned as well as signed 32 bit execution. It is even possible to implement the unsigned 64 bit variant using this.

To get the performance somewhat good I needed the fast-path of the hugeint multiply to be in the header, as otherwise the compiler cant optimize this. I left the slow path in the c++ file, but in case of a modern gcc or clang compiler, calling the multiply function will use the optimized bath and be inlined. This makes the multiply a lot faster (if used directly, not via `*`).

If this looks good I can easily implement this for `uint64` and `(u)int{8/16}`

JAicewizard commented:
Tests that seem to be failing that I cant controll ATM are java, windows, and regression, amalgamation should be fixed by the latest commit

JAicewizard commented:
@Mytherin Could you look at it again this week? Then I can finish the implementation for the other types this weekend

lnkuiper commented on src/common/types/hugeint.cpp:
```diff
@@ -291,8 +292,7 @@ bool Hugeint::TryMultiply(hugeint_t lhs, hugeint_t rhs, hugeint_t &result) {
 // Multiply without overflow check
-template <>
-hugeint_t Hugeint::Multiply<false>(hugeint_t lhs, hugeint_t rhs) {
+hugeint_t Hugeint::SlowMultiply(hugeint_t lhs, hugeint_t rhs) {
```
Why do we need separate `Multiply`/`SlowMultiply` implementations?

JAicewizard replied:
The 2 implementations always existed, but how they are split in 2 functions. The reason this is that if we have the fast version, we want this to be inlined, and thus it must be in the header file. The fast version is very simple, and doesnt take much compile time, even when included a lot of time it should not be notcable.
The slow version is slow, so inlining doesn't help much. It is however very complex, and takes much more time to compile. If this we put this in the header file, this would increase compile times (on compilers that don't support 128 bit), since it has to be compiled every time it is included in another file.

The solution is to compile the slow code once in the .cpp, and, if available, compile the fast version on every include (but since this is very simple code it doesn't affect compile times much).

lnkuiper reviewed (changes requested):
Hi, this PR looks good, I just have one question/concern:

lnkuiper reviewed (approved):
Thanks for the explanation! This is good to go from my side. The PR needs to be pointing at `feature` now instead of `main`, and needs to pull in some changes from `feature` and resolve conflicts.

JAicewizard committed:
(u)hugeint multiply will now be inlined when uint128 is supported

This allows for significantly faster multiplication of (u)hugeint on these compilers

JAicewizard committed:
Implement fast division for 32 both bit integers.

This computes the multiplicative inverse ahead of time, when dividing
multiply by this instead. Multiplication is a lot faster than division.

JAicewizard committed:
Automatically optimise constrhs division to use fast function

JAicewizard committed:
Add support for (u)int{8,16} and uint64 fast constant division

JAicewizard commented:
Rebased, combined commits to be cleaner, and added support for (u)int8/16 as well as uint64.

lnkuiper commented on src/core_functions/scalar/math/numeric.cpp:
```diff
@@ -1323,4 +1332,206 @@ ScalarFunctionSet LeastCommonMultipleFun::GetFunctions() {
+
+static uint16_t ComputeM(uint8_t rhs) {
+	return UINT32_C(0xFFFF) / rhs + 1;
+}
```
Would a template like this make sense so that we can deduplicate code?
```c++
template <class RESULT_TYPE, class RHS_TYPE>
static inline RESULT_TYPE ComputeM(const RHS_TYPE &rhs) {
    return RESULT_TYPE(-1) / rhs + 1;
}
```

lnkuiper commented on src/core_functions/scalar/math/numeric.cpp:
```diff
@@ -1323,4 +1332,206 @@ ScalarFunctionSet LeastCommonMultipleFun::GetFunctions() {
+	// Compute the bottom 32 bits, then the top 32 bits, add them, and only keep the top 32 bits
+	// The reason we need the bottom 32 bits, is because the carry will roll into the top 32 bits.
+	return ((((m & 0xFF) * a) >> 8) + ((m >> 8) * a)) >> 8;
+}
```
I think we can also do more templating here:
```c++
template <class RESULT_TYPE, class OTHER_TYPE>
inline static uint8_t Fastdiv(RESULT_TYPE a, OTHER_TYPE m, RESULT_TYPE d) {
 	return ((((m & RESULT_TYPE(-1)) * a) >> sizeof(RESULT_TYPE) * 8) + ((m >> sizeof(RESULT_TYPE) * 8) * a)) >> sizeof(RESULT_TYPE) * 8;
 }
```
By templating you only have to write the code once, so there's less chance of making mistakes.

Templating usually works for all integral types except `hugeint` and `uhugeint`, as we've created these types ourselves.

lnkuiper commented on src/core_functions/scalar/math/numeric.cpp:
```diff
@@ -1323,4 +1332,206 @@ ScalarFunctionSet LeastCommonMultipleFun::GetFunctions() {
+	if (d < 0) {
+		return -(int8_t)(highbits);
+	}
+	return (int8_t)(highbits);
```
Could you replace the casting here with `UnsafeNumericCast`? This is a helper function that does an efficient static cast with our release builds, but performs over-/underflow checking for debug builds, which helps with finding bugs.

lnkuiper reviewed (changes requested):
Thanks for the changes! It mostly looks good, I just have a few remarks:

JAicewizard committed:
Use templates for FastDiv and ComputeM

JAicewizard committed:
Fix formatting

JAicewizard committed:
fix warning

JAicewizard committed:
try again

JAicewizard committed:
Remove duplicate include

JAicewizard committed:
Rename struct

JAicewizard committed:
Remove more duplicate includes

JAicewizard committed:
??

JAicewizard commented:
@Mytherin @lnkuiper It's been a while, but I re-based and all the test now pass (not sure why they failed before). Are there any more changes I should make?

Mytherin commented:
Thanks for the changes! I think this is actually ready to merge, @lnkuiper thoughts?

lnkuiper reviewed (approved):
@Mytherin I agree, this looks good!
